src/main/java/net/minecraftforge/fmp/block/BlockCoverable.java

```diff
@@ -0,0 +1,730 @@
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.math.AxisAlignedBB;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.RayTraceResult;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.world.IBlockAccess;
+import net.minecraft.world.World;
+import net.minecraft.world.WorldServer;
+import net.minecraftforge.client.event.DrawBlockHighlightEvent;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.common.property.ExtendedBlockState;
+import net.minecraftforge.common.property.IExtendedBlockState;
+import net.minecraftforge.fml.common.eventhandler.EventPriority;
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
+import net.minecraftforge.fmp.ForgeMultipart;
+import net.minecraftforge.fmp.client.multipart.AdvancedParticleManager;
+import net.minecraftforge.fmp.client.multipart.ICustomHighlightPart;
+import net.minecraftforge.fmp.client.multipart.MultipartStateMapper;
+import net.minecraftforge.fmp.microblock.IMicroblockContainerTile;
+import net.minecraftforge.fmp.microblock.MicroblockContainer;
+import net.minecraftforge.fmp.multipart.IMultipart;
+import net.minecraftforge.fmp.multipart.Multipart;
+import net.minecraftforge.fmp.multipart.MultipartContainer;
+import net.minecraftforge.fmp.multipart.MultipartRegistry;
+import net.minecraftforge.fmp.multipart.PartState;
+import net.minecraftforge.fmp.raytrace.RayTraceUtils;
+
+/**
+ * An {@link Block} that allows your block to act as a microblock container.<br/>
+ * Extend this class if you want your block to be able to have microblocks, but not any other kinds of multiparts.
+ * Otherwise, extend {@link Multipart} or make a custom implementation of {@link IMultipart}.<br/>
+ * All the overriden methods have a "default" counterpart that allows you to handle interactions with your block if the
+ * player isn't interacting with a multipart.<br/>
+ * Extend {@link TileCoverable} or implement {@link IMicroblockContainerTile} and return it in
+ * {@link BlockCoverable#createNewTileEntity(World, int)} if you want a custom tile entity for your block.
+ */
+public class BlockCoverable extends Block implements ITileEntityProvider
+{
+    
+    private AxisAlignedBB bounds = FULL_BLOCK_AABB;
+
+    public BlockCoverable(Material material)
+    {
+        super(material);
+        MinecraftForge.EVENT_BUS.register(this);
+    }
+
+    @Override
+    public TileEntity createNewTileEntity(World worldIn, int meta)
+    {
+        return createNewTileEntityDefault(worldIn, meta);
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T extends TileEntity & IMicroblockContainerTile> T createNewTileEntityDefault(World worldIn, int meta)
+    {
+        return (T) new TileCoverable();
+    }
+
+    protected IMicroblockContainerTile getMicroblockTile(IBlockAccess world, BlockPos pos)
+    {
```

Comment:
Black magic using generics. It forced you to return an object that extended TileEntity and implemented IMicroblockContainerTile. I'll get rid of it in favour of capabilities.